src/main/java/com/squareup/kotlinpoet/TypeVariableName.kt

```diff
@@ -65,32 +66,52 @@ class TypeVariableName private constructor(
   }
 
   companion object {
-    internal fun of(name: String, bounds: List<TypeName>): TypeVariableName {
+    internal fun of(name: String, bounds: List<TypeName>, variance: KModifier?): TypeVariableName {
+      require(variance == null || variance == KModifier.IN || variance == KModifier.OUT) {
+        "$variance is an invalid variance modifier, the only allowed values are in and out!"
+      }
       // Strip java.lang.Object from bounds if it is present.
-      return TypeVariableName(name, bounds.filter { it != ANY })
+      return TypeVariableName(name, bounds.filter { it != ANY }, variance)
     }
 
     /** Returns type variable named `name` without bounds.  */
     @JvmStatic @JvmName("get")
-    operator fun invoke(name: String) = TypeVariableName.of(name, emptyList())
+    operator fun invoke(name: String) = TypeVariableName.of(name, emptyList(), variance = null)
+
+    /** Returns type variable named `name` with `variance` and without bounds.  */
+    @JvmStatic @JvmName("get")
+    operator fun invoke(name: String, variance: KModifier) =
+        TypeVariableName.of(name, emptyList(), variance)
 
     /** Returns type variable named `name` with `bounds`.  */
     @JvmStatic @JvmName("get")
-    operator fun invoke(name: String, vararg bounds: TypeName): TypeVariableName {
-      return TypeVariableName.of(name, bounds.toList())
-    }
+    operator fun invoke(name: String, vararg bounds: TypeName) =
+        TypeVariableName.of(name, bounds.toList(), variance = null)
+
+    /** Returns type variable named `name` with `variance` and `bounds`.  */
+    @JvmStatic @JvmName("get")
+    operator fun invoke(name: String, variance: KModifier, vararg bounds: TypeName) =
+        TypeVariableName.of(name, bounds.toList(), variance)
 
     /** Returns type variable named `name` with `bounds`.  */
     @JvmStatic @JvmName("get")
-    operator fun invoke(name: String, vararg bounds: KClass<*>): TypeVariableName {
-      return TypeVariableName.of(name, bounds.map { it.asTypeName() })
-    }
+    operator fun invoke(name: String, vararg bounds: KClass<*>) =
+        TypeVariableName.of(name, bounds.map { it.asTypeName() }, variance = null)
+
+    /** Returns type variable named `name` with `variance` and `bounds`.  */
+    @JvmStatic @JvmName("get")
+    operator fun invoke(name: String, variance: KModifier, vararg bounds: KClass<*>) =
```

Comment:
I like the idea of #157, then we can move all Java-specific stuff outside of the core library.